external/fv3fit/tests/test_bptt.py

```diff
@@ -0,0 +1,272 @@
+import pytest
+import xarray as xr
+import tensorflow as tf
+import numpy as np
+from datetime import timedelta
+from cftime import DatetimeProlepticGregorian as datetime
+from fv3fit.keras._models.recurrent import BPTTModel
+import tempfile
+import fv3fit
+
+
+@pytest.fixture
+def sample_dim_name():
+    return "sample"
+
+
+@pytest.fixture
+def dt():
+    return timedelta(seconds=1)
+
+
+@pytest.fixture
+def train_dataset(sample_dim_name, dt):
+    n_samples = 6
+    n_times = 12
+    n_z = 16
+    t_start = datetime(2020, 1, 1)
+    ds = xr.Dataset(
+        data_vars={
+            "air_temperature": xr.DataArray(
+                np.random.randn(n_samples, n_times, n_z),
+                dims=[sample_dim_name, "time", "z"],
+                attrs={"units": "K"},
+            ),
+            "specific_humidity": xr.DataArray(
+                np.random.randn(n_samples, n_times, n_z),
+                dims=[sample_dim_name, "time", "z"],
+                attrs={"units": "kg/kg"},
+            ),
+            "air_temperature_tendency_due_to_model": xr.DataArray(
+                np.random.randn(n_samples, n_times, n_z),
+                dims=[sample_dim_name, "time", "z"],
+            ),
+            "specific_humidity_tendency_due_to_model": xr.DataArray(
+                np.random.randn(n_samples, n_times, n_z),
+                dims=[sample_dim_name, "time", "z"],
+            ),
+            "air_temperature_tendency_due_to_nudging": xr.DataArray(
+                np.random.randn(n_samples, n_times, n_z),
+                dims=[sample_dim_name, "time", "z"],
+            ),
+            "specific_humidity_tendency_due_to_nudging": xr.DataArray(
+                np.random.randn(n_samples, n_times, n_z),
+                dims=[sample_dim_name, "time", "z"],
+            ),
+            "a": xr.DataArray(
+                np.random.randn(n_samples, n_times), dims=[sample_dim_name, "time"],
+            ),
+            "b": xr.DataArray(
+                np.random.randn(n_samples, n_times), dims=[sample_dim_name, "time"],
+            ),
+        },
+        coords={"time": np.asarray([t_start + i * dt for i in range(n_times)])},
+    )
+    return ds
+
+
+def test_predict_model_gives_tendency_of_train_model(
+    train_dataset, sample_dim_name, dt
+):
+    np.random.seed(0)
+    tf.random.set_seed(1)
+    model = BPTTModel(
+        sample_dim_name,
+        ["a", "b"],
+        n_units=32,
+        n_hidden_layers=4,
+        kernel_regularizer=None,
+        train_batch_size=48,
+        optimizer="adam",
+    )
+    model.build_for(train_dataset)
+
+    train_dataset["air_temperature_tendency_due_to_model"][:] = 0.0
```

Comment:
I'll look into it. I think the fixture gives a new value each time, so no. Otherwise this would cause a lot of test failures, quite often, in many repos.

Reply:
This was refactored to avoid using a fixture, so it should no longer be a risk.